math/ncr_modulo_p.cpp

```diff
@@ -25,71 +25,95 @@ namespace math {
  * implementation.
  */
 namespace ncr_modulo_p {
+
 /**
- * @brief Class which contains all methods required for calculating nCr mod p
+ * @namespace utils
+ * @brief this namespace contains the definitions of the functions called from
+ * the class math::ncr_modulo_p::NCRModuloP
  */
-class NCRModuloP {
- private:
-    std::vector<uint64_t> fac{};  /// stores precomputed factorial(i) % p value
-    uint64_t p = 0;               /// the p from (nCr % p)
-
- public:
-    /** Constructor which precomputes the values of n! % mod from n=0 to size
-     *  and stores them in vector 'fac'
-     *  @params[in] the numbers 'size', 'mod'
-     */
-    NCRModuloP(const uint64_t& size, const uint64_t& mod) {
-        p = mod;
-        fac = std::vector<uint64_t>(size);
-        fac[0] = 1;
-        for (int i = 1; i <= size; i++) {
-            fac[i] = (fac[i - 1] * i) % p;
-        }
+namespace utils {
+/**
+ * @brief finds the values x and y such that a*x + b*y = gcd(a,b)
+ *
+ * @param[in] a the first input of the gcd
+ * @param[in] a the second input of the gcd
+ * @param[out] x the Bézout coefficient of a
+ * @param[out] y the Bézout coefficient of b
+ * @return the gcd of a and b
+ */
+int64_t gcdExtended(const int64_t& a, const int64_t& b, int64_t& x,
```

Comment:
why change unsigned int to signed?

Reply:
Basically to resolve the `clang-tidy` warning [`cppcoreguidelines-narrowing-conversions`](https://clang.llvm.org/extra/clang-tidy/checks/cppcoreguidelines/narrowing-conversions.html) (this was done in 848c7dc8b9e1887b9871f4ac51c3b5fe03031f88, the [log of failing CI](https://github.com/vil02/C-Plus-Plus/actions/runs/5031348020/jobs/9024371606)).

I realize that the whole problem solved by this program is about _positive_ numbers. Re-implementing the whole file to only use `std::uint64_t` is beyond the scope of this PR (it is really not just about writing `static_cast<std::uint64_t>(...)` here and there). Moreover the resulting code would be much more _complex and technical_ (overflows, platform dependent wide types etc.), so the educative value of this example might not be that easy to follow.

Current solution is fine, if the `p` fits into 31 bits. If an extra bit would be needed, it is rather straightforward to change the existing code into a template and use it with some wider or _unbounded_ integer type. The version properly extended to `std::uint64_t` would not have this property.

Thanks for this question - this was actually the harder part part of preparing this PR.

Reply:
Thanks for the detailed explaination